server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1793,31 +1797,41 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            // RepositoryData contains full details for every snapshot
+            final List<SnapshotId> snapshotIdsWithoutAllDetails = repositoryData.getSnapshotIds().stream().filter(
+                snapshotId -> repositoryData.getVersion(snapshotId) == null
+                        || repositoryData.getSnapshotDetails(snapshotId).getStartTimeMillis() == -1
+                        || repositoryData.getSnapshotDetails(snapshotId).getEndTimeMillis() == -1).collect(Collectors.toList());
+            if (snapshotIdsWithoutAllDetails.isEmpty() == false) {
+                final Map<SnapshotId, SnapshotDetails> extraDetailsMap = new ConcurrentHashMap<>();
+                final GroupedActionListener<Void> loadExtraDetailsListener = new GroupedActionListener<>(
                     ActionListener.runAfter(
                         new ActionListener<>() {
                             @Override
                             public void onResponse(Collection<Void> voids) {
-                                logger.info("Successfully loaded all snapshot's version information for {} from snapshot metadata",
+                                logger.info("Successfully loaded all snapshots' detailed information for {} from snapshot metadata",
                                     AllocationService.firstListElementsToCommaDelimitedString(
-                                        snapshotIdsWithoutVersion, SnapshotId::toString, logger.isDebugEnabled()));
+                                        snapshotIdsWithoutAllDetails, SnapshotId::toString, logger.isDebugEnabled()));
                             }
 
                             @Override
                             public void onFailure(Exception e) {
-                                logger.warn("Failure when trying to load missing version information from snapshot metadata", e);
+                                logger.warn("Failure when trying to load missing details from snapshot metadata", e);
                             }
-                        }, () -> filterRepositoryDataStep.onResponse(repositoryData.withVersions(updatedVersionMap))),
-                    snapshotIdsWithoutVersion.size());
-                for (SnapshotId snapshotId : snapshotIdsWithoutVersion) {
-                    threadPool().executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.run(loadAllVersionsListener, () ->
-                        updatedVersionMap.put(snapshotId, getSnapshotInfo(snapshotId).version())));
+                        }, () -> filterRepositoryDataStep.onResponse(repositoryData.withExtraDetails(extraDetailsMap))),
+                        snapshotIdsWithoutAllDetails.size());
+                for (SnapshotId snapshotId : snapshotIdsWithoutAllDetails) {
+                    threadPool().executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.run(loadExtraDetailsListener, () ->
```

Comment:
For master+data nodes this can create a lot of task in the snapshot thread pool. Should we use a queue here with "workers" that pulls the snapshot info to retrieve, similarly to what we do when [uploading snapshot files](https://github.com/elastic/elasticsearch/blob/6e747e6573ee5cec93c3d609459a163f258a3a4c/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java#L2316-L2321)?

Reply:
I think it's probably ok to just queue it all up here. If you get here you want to fix the repo data before anything else IMO, no need to be too clever abut fairness here I think.

Reply:
Yeah, I don't have a strong opinion but wanted to raise this point :)

Reply:
Added a comment about this in af16e79d8111b8d10bfd9b6074d08083d9ab95cd.